File: automation-test-playwright/pages/auth/RegisterCompletePage.ts
import { type Locator, type Page } from "@playwright/test";
import { BasePage } from "../core/BasePage";

export class RegisterCompletePage extends BasePage {
  readonly fullNameInput: Locator;
  readonly usernameInput: Locator;
  readonly passwordInput: Locator;
  readonly confirmPasswordInput: Locator;
  readonly completeButton: Locator;

  /**
   * Initializes this page object.
   */
  constructor(page: Page) {
    super(page);
    this.fullNameInput = this.firstVisible(page.locator('input[name="fullName"]'));
    this.usernameInput = this.firstVisible(page.locator('input[name="username"]'));
    this.passwordInput = this.firstVisible(page.locator('input[name="password"]'));
    this.confirmPasswordInput = this.firstVisible(page.locator('input[name="confirmPassword"]'));
    this.completeButton = this.firstVisible(
      this.anyLocator(
        '[data-testid="register-complete-submit"]',
        'form[action="/auth/register/complete"] button[type="submit"]',
        "button.submit",
        'button:has-text("Tạo tài khoản")'
      )
    );
  }

  /**
   * Opens the page.
   */
  async open(ticket: string, email: string): Promise<void> {
    await this.visit(`/register/complete?ticket=${encodeURIComponent(ticket)}&email=${encodeURIComponent(email)}`);
  }

  async waitForLoaded(email?: string): Promise<void> {
    await this.page.waitForURL(/\/register\/complete/);
    await this.waitForVisible(this.fullNameInput);
    await this.waitForVisible(this.usernameInput);
    await this.waitForVisible(this.completeButton);
    await this.dismissSweetAlertIfPresent();
    if (email) {
      await this.waitForLocatorText(this.page.locator("body"), email);
    }
  }

  async completeRegistration(
    fullName: string,
    username: string,
    password: string,
    confirmPassword = password
  ): Promise<void> {
    await this.dismissSweetAlertIfPresent();
    await this.fullNameInput.fill(fullName);
    await this.usernameInput.fill(username);
    await this.passwordInput.fill(password);
    await this.confirmPasswordInput.fill(confirmPassword);
    await this.completeButton.click();
  }

  async waitForPopupContains(text: string | RegExp): Promise<void> {
    await this.waitForVisible(this.toastPopup());
    await this.waitForLocatorText(this.toastPopup(), text);
  }
}

File: automation-test-selenium/src/main/java/org/suntower/pages/auth/RegisterCompletePage.java
package org.suntower.pages.auth;

import org.openqa.selenium.By;
import org.openqa.selenium.WebDriver;
import org.suntower.core.BasePage;

public class RegisterCompletePage extends BasePage {
  private final By fullNameInput = css("input[name='fullName']");
  private final By usernameInput = css("input[name='username']");
  private final By passwordInput = css("input[name='password']");
  private final By confirmPasswordInput = css("input[name='confirmPassword']");
  private final By completeButton =
      anyCss(
          "[data-testid='register-complete-submit']",
          "form[action='/auth/register/complete'] button[type='submit']",
          "button.submit",
          "button[type='submit']");
  private final By body = css("body");

  public RegisterCompletePage(WebDriver driver) {
    super(driver);
  }

  public void open(String ticket, String email) {
    visit("/register/complete?ticket=" + urlEncode(ticket) + "&email=" + urlEncode(email));
  }

  public void waitForLoaded() {
    waitForLoaded(null);
  }

  public void waitForLoaded(String email) {
    waitForUrlContains("/register/complete");
    waitForVisible(fullNameInput);
    waitForVisible(usernameInput);
    waitForVisible(completeButton);
    dismissSweetAlertIfPresent();
    if (email != null && !email.isBlank()) {
      waitForText(body, email);
    }
  }

  public void completeRegistration(String fullName, String username, String password) {
    completeRegistration(fullName, username, password, password);
  }

  public void completeRegistration(String fullName, String username, String password, String confirmPassword) {
    dismissSweetAlertIfPresent();
    fill(fullNameInput, fullName);
    fill(usernameInput, username);
    fill(passwordInput, password);
    fill(confirmPasswordInput, confirmPassword);
    click(completeButton);
  }

  public void waitForPopupContains(String regex) {
    waitForSweetAlertContainsText(regex);
  }
}
